### PR TITLE
Bugfix/10247 venn label placement

### LINKED ISSUE #10247: Venn diagram - overlap label is not center aligned
The label of the overlap should be centered in the overlap area, but it's not.

image

#### Live demo with steps to reproduce

https://codepen.io/anon/pen/morxjp?&editable=true

#### Affected browser(s)

Chrome

### PR DESCRIPTION
**Pull request by jon-a-nygaard**
Fixed #10247, improved data label positions in venn series.

---
# Description
Fixes scope issue with optimization algorithm causing label position to be worse than the initial position.
Also vertically aligns the label in the middle, and adds width so the label breaks when it overflows the shape. 

# Example(s)
- [Demo of issue](https://highcharts.github.io/highcharts-utils/samples/#gh/master/sample/highcharts/demo/venn-diagram). Especially visible with the label "More Expensive".
- [Demo with fix applied](https://highcharts.github.io/highcharts-utils/samples/#gh/bugfix/10247-venn-label-placement/sample/highcharts/demo/venn-diagram) The label "Will take time to deliver" shows the best benefit of having applied width.

# Related issue(s)
- Closes #10247

### CODE REVIEW

File: samples/unit-tests/series-venn/members/demo.js [hunk 1]
```diff
@@ -667,4 +667,40 @@ QUnit.test('sortByTotalOverlap', function (assert) {
...
+    QUnit.test('getCentroid', assert => {
+        const { getCentroid } = NelderMeadModule;
+        assert.deepEqual(
+            getCentroid([
+                [184.16021264966827, 99.75],
+                [184.16021264966827, 95],
+                [193.3682232821517, 95]
+            ]),
+            [184.16021264966827, 97.375],
+            'should calculate the center point between all the coordinates, except the last'
```
Comment (TorsteinHonsi) [inline samples/unit-tests/series-venn/members/demo.js]: Use upper case first for unit test descriptions. Not an explicit style guide requirement, but all existing tests use uc first.

File: samples/unit-tests/series-venn/members/demo.js [hunk 2]
```diff
@@ -667,4 +667,40 @@ QUnit.test('sortByTotalOverlap', function (assert) {
...
+        const fn = ([x, y]) => -(
+            getMarginFromCircles({ x, y }, internal, external)
+        );
+        assert.deepEqual(
+            nelderMead(
+                fn,
+                [184.16021264966827, 95]
+            ),
+            [140.0000000000064, 174.99997672224276],
+            'should something'
```
Comment (TorsteinHonsi) [inline samples/unit-tests/series-venn/members/demo.js]: As above, plus add a description.

File: js/modules/venn.src.js [hunk 3]
```diff
@@ -434,14 +302,56 @@ var getLabelPosition = function getLabelPosition(internal, external) {
 };
 
 /**
- * Calulates data label positions for a list of relations.
+ * Finds the available width for a label, by taking the label position and
+ * finding the largest distance, which is inside all internal circles, and
+ * outside all external circles.
+ *
+ * @private
+ * @todo Add unit tests
```
Comment (TorsteinHonsi) [inline js/modules/venn.src.js]: Is this still a `@todo`? There are now unit tests below...

Comment (jon-a-nygaard) [reply js/modules/venn.src.js]: It was still to be done, but I took the time to add them now to the PR.

**Comment by jon-a-nygaard** [comment]
Thanks @TorsteinHonsi, I think I have resolved what you pointed out in your comments with my last commits. Please see latest changes.

**Pull request merged**